library/alloc/src/sync.rs

```diff
@@ -1561,35 +1587,27 @@ impl<T: ?Sized, A: Allocator> Arc<T, A> {
     where
         A: Clone,
     {
-        // This Relaxed is OK because we're checking the value in the CAS
-        // below.
-        let mut cur = this.inner().weak.load(Relaxed);
-
         loop {
-            // check if the weak counter is currently "locked"; if so, spin.
-            if cur == usize::MAX {
-                hint::spin_loop();
-                cur = this.inner().weak.load(Relaxed);
-                continue;
-            }
+            let prev = this.inner().weak.fetch_add(ONE_WEAK, Acquire);
+
+            if prev & WEAK_LOCKED == 0 {
+                // Not locked.
 
-            // We can't allow the refcount to increase much past `MAX_REFCOUNT`.
-            assert!(cur <= MAX_REFCOUNT, "{}", INTERNAL_OVERFLOW_ERROR);
-
-            // NOTE: this code currently ignores the possibility of overflow
-            // into usize::MAX; in general both Rc and Arc need to be adjusted
-            // to deal with overflow.
-
-            // Unlike with Clone(), we need this to be an Acquire read to
-            // synchronize with the write coming from `is_unique`, so that the
-            // events prior to that write happen before this read.
-            match this.inner().weak.compare_exchange_weak(cur, cur + 1, Acquire, Relaxed) {
-                Ok(_) => {
-                    // Make sure we do not create a dangling Weak
-                    debug_assert!(!is_dangling(this.ptr.as_ptr()));
-                    return Weak { ptr: this.ptr, alloc: this.alloc.clone() };
+                // We can't allow the refcount to increase much past `MAX_REFCOUNT`.
+                assert!(prev <= MAX_REFCOUNT, "{}", INTERNAL_OVERFLOW_ERROR);
+
+                if prev == 0 {
+                    // This is the first weak pointer, so we set the 'there are weak pointers' flag
+                    // in the strong counter, and also add the implicit weak pointer to the weak
+                    // pointer counter.
+                    this.inner().weak.fetch_add(ONE_WEAK, Relaxed);
+                    this.inner().strong.fetch_or(NO_STRONG_WITH_WEAK, Relaxed);
```

Comment:
This flag setting can be an `add` as it will only be done once.

Reply:
Yeah, but I am guessing that there might be more architectures with an atomic bit set instruction than an atomic add instruction.

Reply:
In the paper, we just used `fetch_add` as it reduces the number of primitives that needed to be wait-free for the algorithm to be wait-free. But in practice, I'm sure you're right that a platform that provides a wait-free `fetch_add` will provide an atomic bit set.